File: solutions/python/0001-two-sum.py
"""
Problem: Two Sum
Link: https://leetcode.com/problems/two-sum/
Difficulty: Easy
Topics: Array, Hash Table
Time: O(n)
Space: O(n)
"""


from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        from collections import defaultdict

        d = defaultdict(int)
        for i in range(len(nums)):
            if (target - nums[i]) in d:
                return [i, d[target - nums[i]]]
            d[nums[i]] = i
        return

File: solutions/javascript/0001-two-sum.js
/**
 * Problem: Two Sum
 * Link: https://leetcode.com/problems/two-sum/
 * Difficulty: Easy
 * Topics: Array, Hash Table
 * Time: O(n)
 * Space: O(n)
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  const seen = new Map();

  for (let index = 0; index < nums.length; index += 1) {
    const complement = target - nums[index];
    if (seen.has(complement)) {
      return [seen.get(complement), index];
    }
    seen.set(nums[index], index);
  }

  return [];
};
